File: frontend/apps/ios/Pantopus/Core/Networking/Models/Wallet/WalletDTOs.swift
//
//  WalletDTOs.swift
//  Pantopus
//
//  DTOs for the read-path wallet endpoints in `backend/routes/wallet.js`.
//  Monetary values are integer cents (the `WalletTransaction.amount` /
//  `Wallet.balance` columns are `bigint`); the view-model formats them into
//  display strings. Field names are preserved verbatim from the route
//  responses via explicit `CodingKeys` (the client does not apply
//  `convertFromSnakeCase`).
//

import Foundation

// MARK: - Balance (GET /api/wallet — backend/routes/wallet.js:55)

/// `GET /api/wallet` envelope.
public struct WalletBalanceResponse: Decodable, Sendable, Hashable {
    public let wallet: Wallet

    public struct Wallet: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        /// Available balance in integer cents.
        public let balance: Int
        public let currency: String
        public let frozen: Bool
        public let lifetimeWithdrawals: Int?
        public let lifetimeReceived: Int?

        private enum CodingKeys: String, CodingKey {
            case id
            case balance
            case currency
            case frozen
            case lifetimeWithdrawals = "lifetime_withdrawals"
            case lifetimeReceived = "lifetime_received"
        }
    }
}

// MARK: - Transactions (GET /api/wallet/transactions — backend/routes/wallet.js:124)

/// `GET /api/wallet/transactions` envelope.
public struct WalletTransactionsResponse: Decodable, Sendable, Hashable {
    public let transactions: [WalletTransactionDTO]
    public let total: Int?
    public let limit: Int?
    public let offset: Int?
}

/// A single `WalletTransaction` row. Only the fields the Recent-activity feed
/// renders are modelled; the decoder ignores the rest (`wallet_id`,
/// `payment_id`, `gig_id`, `metadata`, …).
public struct WalletTransactionDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    /// One of `deposit | withdrawal | gig_income | gig_payment | tip_income |
    /// tip_sent | refund | adjustment | transfer_in | transfer_out |
    /// cancellation_fee`.
    public let type: String
    /// Positive integer cents (the DB enforces `amount > 0`); the sign comes
    /// from `direction`.
    public let amount: Int
    /// `credit | debit` — authoritative, `NOT NULL` with a CHECK constraint on
    /// `WalletTransaction` (backend/database/schema.sql:4526,4541) and returned
    /// by `select('*')` in `walletService.getTransactions`. Optional here only
    /// so an older payload still decodes; callers fall back to the `type`
    /// heuristic when it is absent.
    public let direction: String?
    public let description: String?
    public let currency: String?
    /// One of `completed | pending | failed | reversed`.
    public let status: String
    public let counterpartyId: String?
    public let createdAt: String?

    private enum CodingKeys: String, CodingKey {
        case id
        case type
        case amount
        case direction
        case description
        case currency
        case status
        case counterpartyId = "counterparty_id"
        case createdAt = "created_at"
    }
}

// MARK: - Withdraw (POST /api/wallet/withdraw — backend/routes/wallet.js:84)

/// Body for `POST /api/wallet/withdraw`. `amount` is integer cents (min 100).
/// `idempotencyKey` deduplicates double-taps server-side.
public struct WalletWithdrawRequest: Encodable, Sendable, Hashable {
    public let amount: Int
    public let idempotencyKey: String?

    public init(amount: Int, idempotencyKey: String? = nil) {
        self.amount = amount
        self.idempotencyKey = idempotencyKey
    }
}

/// `POST /api/wallet/withdraw` envelope. The new outbound `WalletTransaction`
/// row is returned so the client can optimistically reflect it; we still
/// refresh from the backend (source of truth).
public struct WalletWithdrawResponse: Decodable, Sendable, Hashable {
    public let success: Bool
    public let transaction: WalletTransactionDTO?
    public let message: String?
}

// MARK: - Pending release (GET /api/wallet/pending-release — backend/routes/wallet.js:160)

/// `GET /api/wallet/pending-release` — escrow breakdown. All values are
/// integer cents / counts.
public struct WalletPendingReleaseResponse: Decodable, Sendable, Hashable {
    public let inReviewCents: Int
    public let releasingSoonCents: Int
    public let totalPendingCents: Int
    public let inReviewCount: Int
    public let releasingSoonCount: Int

    private enum CodingKeys: String, CodingKey {
        case inReviewCents = "in_review_cents"
        case releasingSoonCents = "releasing_soon_cents"
        case totalPendingCents = "total_pending_cents"
        case inReviewCount = "in_review_count"
        case releasingSoonCount = "releasing_soon_count"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/wallet/WalletDtos.kt
package app.pantopus.android.data.api.models.wallet

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Read-path DTOs for `backend/routes/wallet.js`. Monetary values are integer
 * cents (`Wallet.balance` / `WalletTransaction.amount` are `bigint`); the
 * view-model formats them into display strings.
 */

// GET /api/wallet — backend/routes/wallet.js:55
@JsonClass(generateAdapter = true)
data class WalletBalanceResponse(
    val wallet: WalletDto,
)

@JsonClass(generateAdapter = true)
data class WalletDto(
    val id: String,
    /** Available balance in integer cents. */
    val balance: Long,
    val currency: String? = null,
    val frozen: Boolean = false,
    @Json(name = "lifetime_withdrawals") val lifetimeWithdrawals: Long? = null,
    @Json(name = "lifetime_received") val lifetimeReceived: Long? = null,
)

// GET /api/wallet/transactions — backend/routes/wallet.js:124
@JsonClass(generateAdapter = true)
data class WalletTransactionsResponse(
    val transactions: List<WalletTransactionDto> = emptyList(),
    val total: Int? = null,
    val limit: Int? = null,
    val offset: Int? = null,
)

@JsonClass(generateAdapter = true)
data class WalletTransactionDto(
    val id: String,
    /** `deposit | withdrawal | gig_income | gig_payment | tip_income |
     *  tip_sent | refund | adjustment | transfer_in | transfer_out |
     *  cancellation_fee`. */
    val type: String,
    /** Positive integer cents; the sign comes from [direction]. */
    val amount: Long,
    /**
     * `credit | debit` — authoritative. `NOT NULL` with a CHECK constraint on
     * `WalletTransaction` (backend/database/schema.sql:4526,4541) and returned by
     * `select('*')` in `walletService.getTransactions`. Nullable here only so an
     * older payload still decodes; callers fall back to the [type] heuristic.
     */
    val direction: String? = null,
    val description: String? = null,
    val currency: String? = null,
    /** `completed | pending | failed | reversed`. */
    val status: String,
    @Json(name = "counterparty_id") val counterpartyId: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
)

// GET /api/wallet/pending-release — backend/routes/wallet.js:160
@JsonClass(generateAdapter = true)
data class WalletPendingReleaseResponse(
    @Json(name = "in_review_cents") val inReviewCents: Long = 0,
    @Json(name = "releasing_soon_cents") val releasingSoonCents: Long = 0,
    @Json(name = "total_pending_cents") val totalPendingCents: Long = 0,
    @Json(name = "in_review_count") val inReviewCount: Int = 0,
    @Json(name = "releasing_soon_count") val releasingSoonCount: Int = 0,
)

// POST /api/wallet/withdraw — backend/routes/wallet.js:84 (Block 3C)
@JsonClass(generateAdapter = true)
data class WalletWithdrawRequest(
    /** Amount in integer cents (min 100). */
    val amount: Long,
    val idempotencyKey: String? = null,
)

@JsonClass(generateAdapter = true)
data class WalletWithdrawResponse(
    val success: Boolean = false,
    val transaction: WalletTransactionDto? = null,
    val message: String? = null,
)
